File: CsharpSocketServer/Game/Room.cs
using System.Collections.Concurrent;
using System.Numerics;
using api.DTOs;

public enum RoomEventType
{
    BattleReady,
    BattleStart,
    PlayerInput,
    ResultAck,
    Disconnect
}

public struct RoomEvent
{
    public RoomEventType EventType { get; init; }
    public int SessionId { get; init; }
    public object? Payload { get; init; }
}

public enum RoomOutEventType
{
    LoadBattle,
    PlayerSpawn,
    StateUpdate,
    TimeUpdate,
    Attack,
    GameResult,
    EnemyExit,
    CloseRoom
}

public struct RoomOutEvent
{
    public RoomOutEventType EventType { get; init; }
    public int SessionId { get; init; }
    public object Payload { get; init; }
}

public readonly struct Box
{
    public float X { get; }
    public float Y { get; }
    public float HalfWidth  { get; }
    public float HalfHeight { get; }

    public Box(float x, float y, float halfW, float halfH)
    {
        X = x;
        Y = y;
        HalfWidth = halfW;
        HalfHeight = halfH;
    }
}

public sealed class Room
{
    public int RoomId { get; }

    public ClientSession Player1 { get; }
    public ClientSession Player2 { get; }

    public ServerPlayer ServerPlayer1 { get; private set; } = null!;
    public ServerPlayer ServerPlayer2 { get; private set; } = null!;

    private static readonly Vector2 LeftSpawn  = new(-7f, -2.3f);
    private static readonly Vector2 RightSpawn = new( 7f, -2.3f);

    private HashSet<int> _readySet = new();
    private bool _gameStarted;
    private bool _battleStarted;

    private int _waitingAckCount = 2;
    private bool _gameEnded;
    private Dictionary<int, bool> _ackReceivedMap = new();
    public bool Closed { get; private set; }
    private bool _closeQueued;

    public float GameTime { get; private set; } = 100.0f;
    public float StateSendAcc { get; private set; } = 0.0f;
    public float TimeSendAcc { get; private set; } = 0.0f;

    private readonly ConcurrentQueue<RoomEvent> _eventQueue = new();
    private readonly ConcurrentQueue<RoomOutEvent> _outEventQueue = new();
    private const int MAX_OUT_EVENTS_PER_TICK = 20;

    public Room(int roomId, ClientSession p1, ClientSession p2)
    {
        RoomId = roomId;
        Player1 = p1;
        Player2 = p2;
        ServerPlayer1 = new ServerPlayer(Player1.UserId, "LEFT", LeftSpawn);
        ServerPlayer2 = new ServerPlayer(Player2.UserId, "RIGHT", RightSpawn);
    }

    public void EnqueueEvent(RoomEvent ev)
    {
        _eventQueue.Enqueue(ev);

        while (_eventQueue.Count > 20)    
            _eventQueue.TryDequeue(out _);
    }

    public void EmitOutEvent(RoomOutEvent ev)
    {
        _outEventQueue.Enqueue(ev);
    }

    private void InputEvents()
    {
        while(_eventQueue.TryDequeue(out var ev))
        {
            switch (ev.EventType)
            {
                case RoomEventType.BattleReady:
                    CheckMatch(ev);
                    break;

                case RoomEventType.BattleStart:
                    PlayerSpawn();
                    break;

                case RoomEventType.PlayerInput:
                    OnInput(ev);
                    break;

                case RoomEventType.ResultAck:
                    OnAckReceived(ev);
                    break;

                case RoomEventType.Disconnect:
                    OnPlayerDisconnect(ev);
                    break;

                default:
                    break;
            }
        }
    }

    private void OutEvents()
    {
        int processed = 0;

        while (processed < MAX_OUT_EVENTS_PER_TICK &&
        _outEventQueue.TryDequeue(out var ev))
        {
            Transport.Dispatch(ev);
            processed++;
        }
    }

    public void Update(float dt)
    {
        InputEvents();
        OutEvents();

        if (!_gameStarted) return;   // 매치 안됐으면 return
        if (Player1 == null || 
        Player2 == null) return;     // 세션 1명이라도 없으면 return
        if (!_battleStarted) return; // 배틀씬 아니면 return

        GameTime -= dt;

        ServerPlayerUpdate(dt);

        StateSendAcc += dt;
        if (StateSendAcc >= 0.033)
            EmitStateUpdate();

        TimeSendAcc += dt;
        if (TimeSendAcc >= 1.0)
            EmitTimeUpdate();

        EmitDamageUpdate();

        if (GameTime <= 0f || 
        ServerPlayer1.CurrentHp <= 0 || 
        ServerPlayer2.CurrentHp <= 0)
        {
            EndGame();
        }
    }

    private void ServerPlayerUpdate(float dt) 
    {
        ServerPlayer1.Update(dt);
        ServerPlayer2.Update(dt);
    }

    void CheckMatch(RoomEvent ev)
    {
        _readySet.Add(ev.SessionId);

        if (_readySet.Count < 2)
            return;

        _gameStarted = true;

        RoomOutEvent p1 = new RoomOutEvent
        {
            EventType = RoomOutEventType.LoadBattle,
            SessionId = Player1.SessionId,
            Payload = new BasePacket { Type = "LOAD_BATTLE" }
        };

        RoomOutEvent p2 = new RoomOutEvent
        {
            EventType = RoomOutEventType.LoadBattle,
            SessionId = Player2.SessionId,
            Payload = new BasePacket { Type = "LOAD_BATTLE" }
        };

        EmitOutEvent(p1);
        EmitOutEvent(p2);
    }

    private void PlayerSpawn()
    {
        if (_battleStarted) return;
        _battleStarted = true;

        var p1State = ServerPlayer1.StatePacket();
        var p2State = ServerPlayer2.StatePacket();

        RoomOutEvent p1 = new RoomOutEvent
        {
            EventType = RoomOutEventType.PlayerSpawn,
            SessionId = Player1.SessionId,
            Payload = p1State
        };

        RoomOutEvent p1p2 = new RoomOutEvent
        {
            EventType = RoomOutEventType.PlayerSpawn,
            SessionId = Player1.SessionId,
            Payload = p2State
        };

        RoomOutEvent p2 = new RoomOutEvent
        {
            EventType = RoomOutEventType.PlayerSpawn,
            SessionId = Player2.SessionId,
            Payload = p1State
        };

        RoomOutEvent p2p1 = new RoomOutEvent
        {
            EventType = RoomOutEventType.PlayerSpawn,
            SessionId = Player2.SessionId,
            Payload = p2State
        };

        EmitOutEvent(p1);
        EmitOutEvent(p1p2);
        EmitOutEvent(p2);
        EmitOutEvent(p2p1);
    }

    private void OnInput(RoomEvent ev)
    {
        if (ev.SessionId == Player1.SessionId)
            ServerPlayer1.ApplyInput((PlayerInputPacket)ev.Payload!);
        else
            ServerPlayer2.ApplyInput((PlayerInputPacket)ev.Payload!);
    }

    private void EmitStateUpdate()
    {
        StateSendAcc = 0.0f;

        if (!ServerPlayer1.IsStateDirty && !ServerPlayer2.IsStateDirty)
            return;

        var p1State = ServerPlayer1.StatePacket();
        var p2State = ServerPlayer2.StatePacket();

        RoomOutEvent p1 = new RoomOutEvent
        {
            EventType = RoomOutEventType.StateUpdate,
            SessionId = Player1.SessionId,
            Payload = p1State
        };

        RoomOutEvent p1p2 = new RoomOutEvent
        {
            EventType = RoomOutEventType.StateUpdate,
            SessionId = Player1.SessionId,
            Payload = p2State
        };

        RoomOutEvent p2 = new RoomOutEvent
        {
            EventType = RoomOutEventType.StateUpdate,
            SessionId = Player2.SessionId,
            Payload = p1State
        };

        RoomOutEvent p2p1 = new RoomOutEvent
        {
            EventType = RoomOutEventType.StateUpdate,
            SessionId = Player2.SessionId,
            Payload = p2State
        };

        EmitOutEvent(p1);
        EmitOutEvent(p1p2);
        EmitOutEvent(p2);
        EmitOutEvent(p2p1);

        ServerPlayer1.ClearStateDirty();
        ServerPlayer2.ClearStateDirty();
    }

    private void EmitDamageUpdate() 
    {
        bool p1Hit = CheckDamage(ServerPlayer1, ServerPlayer2);
        bool p2Hit = CheckDamage(ServerPlayer2, ServerPlayer1);

        if (p1Hit) BroadcastDamage(ServerPlayer2);
        if (p2Hit) BroadcastDamage(ServerPlayer1);
    }

    private void BroadcastDamage(ServerPlayer hurt)
    {
        EmitOutEvent(new RoomOutEvent {
            EventType = RoomOutEventType.Attack,
            SessionId = Player1.SessionId,
            Payload = hurt.HurtPacket()
        });

        EmitOutEvent(new RoomOutEvent {
            EventType = RoomOutEventType.Attack,
            SessionId = Player2.SessionId,
            Payload = hurt.HurtPacket()
        });
    }

    private bool CheckDamage(ServerPlayer atk, ServerPlayer tar)
    {
        // Punch 상태가 아니면 데미지 판정 안 함
        if (atk.State != PlayerState.Punch)
            return false;

        // 이미 판정이 체크된 상태면 데미지 판정 안 함
        if (atk.PunchChecked)
            return false;

        // 펀치 범위에 상대가 없으면 데미지 판정 안 함
        if (!IsInPunchRange(atk, tar))
            return false;

        atk.SetPunchChecked(true);

        // 피격 방향 설정
        if (atk.Position.X > tar.Position.X && tar.Direction < 0)
            tar.SetDir(1);
        else if (atk.Position.X < tar.Position.X && tar.Direction > 0)
            tar.SetDir(-1);

        tar.TakeDamage(10, atk.Direction * 1.0f);

        return true;
    }

    private bool IsInPunchRange(ServerPlayer atk, ServerPlayer tar)
    {
        Box hitbox = new Box
        (
            atk.Position.X + (0.3f * atk.Direction), 
            atk.Position.Y,
            0.3f,
            0.7f
        );

        Box hurtbox = new Box
        (
            tar.Position.X,
            tar.Position.Y + 0.5f,
            0.5f,
            0.5f
        );
        
        return Overlap(hitbox, hurtbox);
    }

    private bool Overlap(Box hit, Box hurt)
    {
        return !(Math.Abs(hit.X - hurt.X) > (hit.HalfWidth + hurt.HalfWidth) ||
                Math.Abs(hit.Y - hurt.Y) > (hit.HalfHeight + hurt.HalfHeight));
    }

    private void EmitTimeUpdate()
    {
        TimeSendAcc -= 1.0f;

        TimeSyncPacket time = new TimeSyncPacket
        {
            Type = "GAME_TIME",
            Time = (int)Math.Ceiling(GameTime)
        };

        RoomOutEvent p1 = new RoomOutEvent
        {
            EventType = RoomOutEventType.TimeUpdate,
            SessionId = Player1.SessionId,
            Payload = time
        };

        RoomOutEvent p2 = new RoomOutEvent
        {
            EventType = RoomOutEventType.TimeUpdate,
            SessionId = Player2.SessionId,
            Payload = time
        };
        
        EmitOutEvent(p1);
        EmitOutEvent(p2);
    }

    public async Task SaveRecordAsync(ClientSession winner, ClientSession loser)
    {
        var req = new SaveRecordRequest
        {
            WinnerId = winner.UserId,
            LoserId = loser.UserId
        };

        bool success = await ApiClient.Post("battle/save", req);

        if (success)
            Console.WriteLine("전적 저장됨");
        else
            Console.WriteLine("전적 저장 실패");
    }

    private void EndGame()
    {
        if (_gameEnded)
            return;

        _gameEnded = true;

        EmitGameResult();
        BeginCloseAckPhase();
    }

    public void EmitGameResult()
    {
        if (ServerPlayer1.CurrentHp > ServerPlayer2.CurrentHp)
        {
            RoomOutEvent p1 = new RoomOutEvent
            {
                EventType = RoomOutEventType.GameResult,
                SessionId = Player1.SessionId,
                Payload = new BasePacket { Type = "GAME_WIN" }
            };

            RoomOutEvent p2 = new RoomOutEvent
            {
                EventType = RoomOutEventType.GameResult,
                SessionId = Player2.SessionId,
                Payload = new BasePacket { Type = "GAME_LOSE" }
            };

            EmitOutEvent(p1);
            EmitOutEvent(p2);

            _ = SaveRecordAsync(Player1, Player2);
        }
        else if (ServerPlayer1.CurrentHp < ServerPlayer2.CurrentHp)
        {
            RoomOutEvent p1 = new RoomOutEvent
            {
                EventType = RoomOutEventType.GameResult,
                SessionId = Player1.SessionId,
                Payload = new BasePacket { Type = "GAME_LOSE" }
            };

            RoomOutEvent p2 = new RoomOutEvent
            {
                EventType = RoomOutEventType.GameResult,
                SessionId = Player2.SessionId,
                Payload = new BasePacket { Type = "GAME_WIN" }
            };

            EmitOutEvent(p1);
            EmitOutEvent(p2);

            _ = SaveRecordAsync(Player2, Player1);
        }
        else
        {
            RoomOutEvent p1 = new RoomOutEvent
            {
                EventType = RoomOutEventType.GameResult,
                SessionId = Player1.SessionId,
                Payload = new BasePacket { Type = "GAME_DRAW" }
            };

            RoomOutEvent p2 = new RoomOutEvent
            {
                EventType = RoomOutEventType.GameResult,
                SessionId = Player2.SessionId,
                Payload = new BasePacket { Type = "GAME_DRAW" }
            };

            EmitOutEvent(p1);
            EmitOutEvent(p2);

            _ = SaveRecordAsync(Player1, Player2);
        }
    }

    private void BeginCloseAckPhase()
    {
        _ackReceivedMap.Clear();
        _ackReceivedMap[Player1.SessionId] = false;
        _ackReceivedMap[Player2.SessionId] = false;

        _waitingAckCount = 2;

        RoomOutEvent p1 = new RoomOutEvent
        {
            EventType = RoomOutEventType.GameResult,
            SessionId = Player1.SessionId,
            Payload = new BasePacket { Type = "ROOM_CLOSED" }
        };

        RoomOutEvent p2 = new RoomOutEvent
        {
            EventType = RoomOutEventType.GameResult,
            SessionId = Player2.SessionId,
            Payload = new BasePacket { Type = "ROOM_CLOSED" }
        };

        EmitOutEvent(p1);
        EmitOutEvent(p2);
    }

    private void OnAckReceived(RoomEvent re)
    {
        if (_ackReceivedMap[re.SessionId]) return;

        _ackReceivedMap[re.SessionId] = true;
        _waitingAckCount--;

        if (_waitingAckCount == 0)
            Closed = true;
    }

    private void OnPlayerDisconnect(RoomEvent re)
    {
        if (Closed) return;
        Closed = true;

        ClientSession? winner = null;
        ClientSession? loser = null;

        if (re.SessionId == Player1.SessionId) {
            winner = Player2;
            loser = Player1;
        }
        else if (re.SessionId == Player2.SessionId) {
            winner = Player1;
            loser = Player2;
        }
        else return;

        RoomOutEvent exit = new RoomOutEvent
        {
            EventType = RoomOutEventType.EnemyExit,
            SessionId = winner.SessionId,
            Payload = new BasePacket { Type = "ENEMY_EXIT" }
        };

        RoomOutEvent close = new RoomOutEvent
        {
            EventType = RoomOutEventType.CloseRoom,
            SessionId = winner.SessionId,
            Payload = new BasePacket { Type = "ROOM_CLOSED" }
        };

        EmitOutEvent(exit);
        EmitOutEvent(close);
        _ = SaveRecordAsync(winner, loser);
    }

    public bool TryQueueClose()
    {
        if (!Closed) return false;
        if (_closeQueued) return false;

        _closeQueued = true;
        return true;
    }

    public void CloseRoom()
    {
        if (Closed) return;

        Player1.Room = null;
        Player2.Room = null;
    }
}

File: CppSocketServer/CppSocketServer/Room.cpp
#include <memory>
#include <cmath>
#include <iostream>
#include <nlohmann/json.hpp>
#include "ClientSession.h"
#include "Room.h"
#include "Transport.h"
#include "ServerPlayer.h"
#include "ApiClient.h"
#include "SaveRecordRequest.h"
#include "MatchFoundPacket.h"
#include "RoomEvent.h"
#include "ThreadPool.h"
#include "PlayerStruct.h"
#include "Server.h"

Room::Room(int id, int player1,int player2, ThreadPool& pool)
    : roomId(id), p1sid(player1), p2sid(player2), threadPool(pool)
{
    Vector2 leftSpawn = { -7.0f, -2.5f };
    Vector2 rightSpawn = { 7.0f, -2.5f };

    sp1 = std::make_unique<ServerPlayer>(
        p1UserId,
        static_cast<int8_t>(Side::Left),
        leftSpawn
    );

    sp2 = std::make_unique<ServerPlayer>(
        p2UserId,
        static_cast<int8_t>(Side::Right),
        rightSpawn
    );
}

// Network/IO 스레드 → Room 스레드로 전달되는 이벤트 enqueue
void Room::EnqueueEvent(const RoomEvent& ev)
{
    {
        std::lock_guard<std::mutex> lock(eventMutex);
        eventQueue.push(ev);
    }
    
    // 이벤트 큐 원소가 50개 이상이면 틱 밀리는 룸 로그 띄우기
    size_t sz;
    {
        std::lock_guard<std::mutex> lock(eventMutex);
        sz = eventQueue.size();
    }

    if (sz > 50)
    {
        auto now = std::chrono::steady_clock::now();

        if (now - lastLog > std::chrono::seconds(1))
        {
            std::cout
                << "[ROOM EVENT BACKLOG] room=" << roomId
                << " size=" << sz << "\n";
            lastLog = now;
        }
    }
}

// Network/IO 스레드로 전달되는 이벤트 enqueue
void Room::EmitOutEvent(const RoomOutEvent& ev)
{
    {
        std::lock_guard<std::mutex> lock(outEventMutex);
        outEventQueue.push(ev);
    }
    
    size_t sz;
    {
        std::lock_guard<std::mutex> lock(outEventMutex);
        sz = outEventQueue.size();
    }

    if (sz > 20)
    {
        auto now = std::chrono::steady_clock::now();

        if (now - lastLog > std::chrono::seconds(1))
        {
            std::cout
                << "[ROOM OUT BACKLOG] room=" << roomId
                << " size=" << sz << "\n";
            lastLog = now;
        }
    }
}

// 이벤트 큐 처리 (룸 업데이트 스레드에서 호출)
void Room::InputEvents()
{
    {
        std::lock_guard<std::mutex> lock(eventMutex);

        while (!eventQueue.empty())
        {
            auto ev = eventQueue.front();
            eventQueue.pop();

            switch (ev.type)
            {
            case RoomEventType::BattleReady:
                CheckMatch(ev);
                break;

            case RoomEventType::BattleStart:
                PlayerSpawn(ev);
                break;

            case RoomEventType::PlayerInput:
                OnInput(ev);
                break;

            case RoomEventType::ResultAck:
                OnAckReceived(ev);
                break;

            case RoomEventType::Disconnect:
                OnPlayerDisconnect(ev);
                break;

            default:
                break;
            }
        }
    }
}

// 이벤트 큐 처리 (룸 업데이트 스레드에서 호출)
void Room::OutEvents()
{
    {
        std::lock_guard<std::mutex> lock(outEventMutex);

        while (!outEventQueue.empty())
        {
            auto ev = outEventQueue.front();
            outEventQueue.pop();
            net.Dispatch(ev);
        }
    }
}

void Room::Update(double dt)
{
    InputEvents();
    OutEvents();

    if (!gameStarted) return;   // 매치 안됐으면 return
    if (!Server::Instance().FindSession(p1sid) ||
        !Server::Instance().FindSession(p2sid)) return;     // 세션 1명이라도 없으면 return
    if (!battleStarted) return; // 배틀씬 아니면 return

    gameTime -= dt;

    ServerPlayerUpdate(dt);

    stateSendAcc += dt;
    if (stateSendAcc >= 0.033)
        EmitStateUpdate();

    timeSendAcc += dt;
    if (timeSendAcc >= 1.0f)
        EmitTimeUpdate();

    EmitDamageUpdate();

    if (gameTime <= 0.0f || sp1->GetCurrentHP() <= 0 || sp2->GetCurrentHP() <= 0)
        EndGame();
}

void Room::CheckMatch(const RoomEvent& re)
{
    if (gameStarted) return;

    ready.insert(re.sessionId);

    if (ready.size() < 2) return;

    gameStarted = true;

    EmitOutEvent({ RoomOutEventType::LoadBattle, p1sid });
    EmitOutEvent({ RoomOutEventType::LoadBattle, p2sid });
}

void Room::PlayerSpawn(const RoomEvent& re)
{
    battleStarted = true;

    auto s1 = sp1->StatePacket();
    auto s2 = sp2->StatePacket();

    EmitOutEvent({ RoomOutEventType::PlayerSpawn, p1sid, UpdateStatePayload{ s1, s2 } });
    EmitOutEvent({ RoomOutEventType::PlayerSpawn, p2sid, UpdateStatePayload{ s1, s2 } });
}

// Player로직에 Input 값 넘겨주기
void Room::OnInput(const RoomEvent& re)
{
    if (re.sessionId == p1sid)
        sp1->ApplyInput(re.input);
    else if (re.sessionId == p2sid)
        sp2->ApplyInput(re.input);
}

void Room::ServerPlayerUpdate(double dt) 
{
    sp1->Update(dt);
    sp2->Update(dt);
}

void Room::EmitStateUpdate()
{
    stateSendAcc = 0.0f;

    if (!sp1->IsStateDirty() && !sp2->IsStateDirty())
        return;

    auto s1 = sp1->StatePacket();
    auto s2 = sp2->StatePacket();

    EmitOutEvent({ RoomOutEventType::StateUpdate, p1sid, UpdateStatePayload{ s1, s2 } });
    EmitOutEvent({ RoomOutEventType::StateUpdate, p2sid, UpdateStatePayload{ s1, s2 } });

    sp1->ClearStateDirty();
    sp2->ClearStateDirty();
}

void Room::EmitTimeUpdate()
{
    timeSendAcc -= 1.0f;

    auto t = static_cast<int32_t>(std::ceil(gameTime));

    EmitOutEvent({ RoomOutEventType::TimeUpdate, p1sid, UpdateTimePayload {t} });
    EmitOutEvent({ RoomOutEventType::TimeUpdate, p2sid, UpdateTimePayload {t} });
}

void Room::EmitDamageUpdate() {

    if (CheckDamage(*sp1, *sp2)) 
    {
        auto d2 = sp2->HurtPacket();
        EmitOutEvent({ RoomOutEventType::Attack, p1sid, UpdateHurtPayload { d2 } });
        EmitOutEvent({ RoomOutEventType::Attack, p2sid, UpdateHurtPayload { d2 } });
    }
    else if (CheckDamage(*sp2, *sp1))
    {
        auto d1 = sp1->HurtPacket();
        EmitOutEvent({ RoomOutEventType::Attack, p1sid, UpdateHurtPayload { d1 } });
		EmitOutEvent({ RoomOutEventType::Attack, p2sid, UpdateHurtPayload { d1 } });
    }
}

bool Room::CheckDamage(ServerPlayer& attacker, ServerPlayer& target)
{
	// Punch 상태가 아니면 데미지 판정 안 함
    if (attacker.GetState() != PlayerState::Punch)
        return false;

	// 이미 판정이 체크된 상태면 데미지 판정 안 함
    if (attacker.HasPunchChecked())
        return false;

    attacker.SetPunchChecked(true);

	// 펀치 범위에 상대가 없으면 데미지 판정 안 함
    if (!IsInPunchRange(attacker, target))
        return false;

	// 피격 방향 설정
    if (attacker.GetPosition().x > target.GetPosition().x && target.GetDir() < 0)
        target.SetDir(1);
    else if (attacker.GetPosition().x < target.GetPosition().x && target.GetDir() > 0)
        target.SetDir(-1);

    target.TakeDamage(10, attacker.GetDir() * 1.0f);

    return true;
}

// hitbox와 hurtbox 위치 및 크기 생성
bool Room::IsInPunchRange(ServerPlayer& attacker, ServerPlayer& target)
{
    Hitbox hitbox;
    Hurtbox hurtbox;

    hitbox.x = attacker.GetPosition().x + (0.3f * attacker.GetDir());
    hitbox.y = attacker.GetPosition().y;
    hitbox.halfW = 0.3f;
    hitbox.halfH = 0.7f;

    hurtbox.x = target.GetPosition().x;
    hurtbox.y = target.GetPosition().y + 0.5f;
    hurtbox.halfW = 0.5f;
    hurtbox.halfH = 0.5f;

    return Overlap(hitbox, hurtbox);
}

// hitbox와 hurtbox 겹침 판정
bool Room::Overlap(Hitbox& hit, Hurtbox& hurt)
{
    float diffX = std::abs(hit.x - hurt.x);
    float diffY = std::abs(hit.y - hurt.y);
    float limitX = hit.halfW + hurt.halfW;
    float limitY = hit.halfH + hurt.halfH;

    return (diffX <= limitX) && (diffY <= limitY);
}

bool Room::IsBotUser(int userId)
{
    return userId <= 0;
}

void Room::EndGame()
{
    if (closed.exchange(true))
        return;

    EmitGameResult();
    BeginCloseAckPhase();
}

void Room::EmitGameResult()
{
    int p1hp = sp1->GetCurrentHP();
    int p2hp = sp2->GetCurrentHP();

    if (p1hp > p2hp)
    {
        EmitOutEvent({ RoomOutEventType::GameResult, p1sid, GameResultPayload { p1sid } });
        EmitOutEvent({ RoomOutEventType::GameResult, p2sid, GameResultPayload { p1sid } });
        SaveRecordAsync(p1UserId, p2UserId);
    }
    else if (p1hp < p2hp)
    {
        EmitOutEvent({ RoomOutEventType::GameResult, p1sid, GameResultPayload { p2sid } });
        EmitOutEvent({ RoomOutEventType::GameResult, p2sid, GameResultPayload { p2sid } });
        SaveRecordAsync(p2UserId, p1UserId);
    }
    else
    {
        EmitOutEvent({ RoomOutEventType::GameResult, p1sid, GameResultPayload { -1 } });
        EmitOutEvent({ RoomOutEventType::GameResult, p2sid, GameResultPayload { -1 } });
        SaveRecordAsync(p1UserId, p2UserId);
    }
}

// 게임 종료 후 ACK 패킷 대기 단계 시작
void Room::BeginCloseAckPhase()
{
    ackReceivedMap.clear();
    ackReceivedMap[p1sid] = false;
    ackReceivedMap[p2sid] = false;

    waitingAckCount = 2;

    EmitOutEvent({ RoomOutEventType::CloseRoom, p1sid });
    EmitOutEvent({ RoomOutEventType::CloseRoom, p2sid });
}

// 클라이언트에서 ACK 패킷 수신 시 종료 플래그 처리
void Room::OnAckReceived(const RoomEvent& re)
{
    int sid = re.sessionId;
    auto it = ackReceivedMap.find(sid);

    if (it == ackReceivedMap.end())
        return; // 이 ACK 페이즈 대상이 아님

    if (it->second)
        return;

	it->second = true;
    waitingAckCount--;

    if (waitingAckCount == 0)
		closed = true;
}

// 한 쪽이 강제종료 or 연결 끊김 시 처리
void Room::OnPlayerDisconnect(const RoomEvent& re)
{
    if (closed.exchange(true))
        return;

    int winnerUserId = -1;
    int loserUserId = -1;

    if (re.sessionId == p1sid) {
        winnerUserId = p2UserId;
        loserUserId = p1UserId;

        EmitOutEvent({ RoomOutEventType::EnemyExit, p1sid });
        EmitOutEvent({ RoomOutEventType::CloseRoom, p1sid });
        SaveRecordAsync(winnerUserId, loserUserId);
    }
    else if (re.sessionId == p2sid) {
        winnerUserId = p1UserId;
        loserUserId = p2UserId;

        EmitOutEvent({ RoomOutEventType::EnemyExit, p2sid });
        EmitOutEvent({ RoomOutEventType::CloseRoom, p2sid });
        SaveRecordAsync(winnerUserId, loserUserId);
    }
    else return;
}

// ThreadPool을 이용해 비동기 전적 저장
void Room::SaveRecordAsync(int winner, int loser)
{
    if (IsBotUser(winner) || IsBotUser(loser))
    {
        std::cout << "[Battle] bot match, skip record\n";
        return;
    }

    SaveRecordRequest req{ winner, loser };

    // Json 직렬화 + API 요청은 비용이 크기 때문에 Thread Pool 사용
    threadPool.Enqueue([req]() mutable {
        try
        {
            nlohmann::json j = req;
            std::string body = j.dump();

            bool success = ApiClient::Instance().Post("battle/save", body);

            if (success)
                std::cout << "[Battle] 전적 저장 성공\n";
            else
                std::cout << "[Battle] 전적 저장 실패\n";
        }
        catch (const std::exception& e)
        {
            std::cerr << "[Battle] SaveRecord exception: " << e.what() << "\n";
        }
        catch (...)
        {
            std::cerr << "[Battle] SaveRecord unknown exception\n";
        }
    });
}

void Room::CloseRoom()
{
    if (closed.exchange(true))
        return;

	ClientSession* s1 = Server::Instance().FindSession(p1sid);
    ClientSession* s2 = Server::Instance().FindSession(p2sid);

    if (s1) s1->SetRoomId(-1);
    if (s2) s2->SetRoomId(-1);
}

